[Speed] Adds Speculation docs (#16587)
cloudflare/cloudflare-docs: src/content/docs/speed/optimization/content/speculation.mdx

---
pcx_content_type: how-to
title: Speculation
sidebar: 
  order: 9
---

import { FeatureTable, TabItem, Tabs } from "~/components";

Speculation is a tool for improving web page performance by prefetching the most likely next navigation.

---

## Availability

<FeatureTable id="speed.speculation" />

---

## Requirements

Speculation works under the following conditions:
- The Speculation feature is enabled in Cloudflare.

> - The browser of the web page visitor is using a Chromium-based browser v121 or later.

Reviewer: ```suggestion
- The browser of the web page visitor is using a Chromium-based browser version 121 or later.
```

> - The web page requested by the prefetch is eligible for cache.

Reviewer 2: Should we point here to the documentation on how to cache HTML?
https://developers.cloudflare.com/cache/how-to/cache-rules/examples/cache-everything/

## How Speculation works

When Cloudflare’s Speculation feature is enabled, an HTTP header called `Speculation-Rules` is added to web page responses. The value for this header is an URL that hosts an opinionated Speculation-Rules configuration. This configuration instructs the browser to consider prefetching any future navigations with a `conservative` [eagerness](https://developer.chrome.com/docs/web-platform/prerender-pages#eagerness).

The configuration looks like this:

```json
{
  "prefetch": [{
    "source": "document",
    "where": {
      "and": [
        { "href_matches": "/*", "relative_to": "document" },
      ]
    },
    "eagerness": "conservative"
  }]
}
```

This configuration instructs the browser to initiate prefetch requests for future navigations. These prefetch requests will include the `sec-purpose: prefetch` HTTP request header. Prefetches that are not successful will respond with a `503` status code. Prefetches that are successful will respond with a `200` status code.

## Test Speculation

> To test that Speculation is enabled, you can check that your HTTP response headers for your web pages include the `Speculation-Rules` header. However, note that during the Beta phase of Speculation, this behavior might not be 100% consistent.

Reviewer: You mention `beta phase` here. Does this mean that this needs a beta tag in the title of the page?

> To test that Speculation is enabled, you can check that your HTTP response headers for your web pages include the `Speculation-Rules` header. However, note that during the Beta phase of Speculation, this behavior might not be 100% consistent.

Reviewer: ```suggestion
To test that Speculation is enabled, you can check that your HTTP response headers for your web pages include the `Speculation-Rules` header. However, note that during the beta phase of Speculation, this behavior might not be 100% consistent.
```

> To test whether your browser is making prefetch requests, open the **Network** tab in Chrome DevTools. Then, mouse-down on a link on a webpage with Speculation enabled. This action should initiate a prefetch request, which will be  visible in the **Network** tab. However, note that there are several reasons why the browser might choose not to initiate a prefetch. Refer to the [Chrome Limits guide](https://developer.chrome.com/docs/web-platform/prerender-pages#chrome-limits) for more details. For more general information about debugging Speculation, refer to the [Chrome Speculation Debugging guide](https://developer.chrome.com/docs/devtools/application/debugging-speculation-rules).

Reviewer: ```suggestion
To test whether your browser is making prefetch requests, open the **Network** tab in Chrome DevTools. Then, select a link on a webpage with Speculation enabled. This action should initiate a prefetch request, which will be  visible in the **Network** tab. However, note that there are several reasons why the browser might choose not to initiate a prefetch. Refer to the [Chrome Limits guide](https://developer.chrome.com/docs/web-platform/prerender-pages#chrome-limits) for more details. For more general information about debugging Speculation, refer to the [Chrome Speculation Debugging guide](https://developer.chrome.com/docs/devtools/application/debugging-speculation-rules).
```

## RUM integration

> Speculation is designed to integrate with Web Analytics & Real User Measurements (RUM). This integration allows you to understand the web performance implications of Speculation within the Web Analytics UI in Cloudflare’s Dashboard.

Reviewer: ```suggestion
Speculation is designed to integrate with Web Analytics & Real User Measurements (RUM). This integration allows you to understand the web performance implications of Speculation within the Web Analytics interface in Cloudflare’s Dashboard.
```

While you can use Speculation without RUM enabled, you will not have visibility into how the feature is affecting the performance of your web pages. For further details on how to set up RUM, refer to the [Web Analytics & RUM](/web-analytics/) documentation.

## Enable and disable Speculation

Speculation is available in Cloudflare’s **Speed** tab of the dashboard and also in the API and Terraform.

<Tabs syncKey="dashPlusAPI"> <TabItem label="Dashboard">

To enable **Prefetch URLs** in the dashboard:

1. Log in to your [Cloudflare account](https://dash.cloudflare.com), select your account and go to a specific domain.
2. Go to **Speed** > **Optimization** > **Content Optimization**.

> 3. For **Speculation**, switch the toggle to **On**.

Reviewer: ```suggestion
3. Enable **Speculation**.
```

</TabItem> <TabItem label="API">

Use the following `PATCH` request to enable Speculation:

```bash
curl --request PATCH \
"https://api.cloudflare.com/client/v4/zones/{zone_id}/settings/speculation" \
--header "Authorization: Bearer <API_TOKEN>" \
--header "Content-Type: application/json" \
--data '{
  "value":"on"
  }
```

> To disable Speculation, set value to "off".

Reviewer: ```suggestion
To disable Speculation, set `value:` to `"off"`.
```
